### PR TITLE
bpo-40939: Rename PyPegen* functions to PyParser*

### PR DESCRIPTION
**Pull request by lysnikolaou**
Renaame PyPegen* functions to PyParser*, so that we can remove the
old set of PyParser* functions that were using the old parser, but keep
everything backwards-compatible.

https://bugs.python.org/issue40939

### CODE REVIEW

**Comment by pablogsal** [comment]
Seems that there some errors in Windows in Azure

**Comment by pablogsal** [comment]
The windows build still fails, not sure where the errors are coming from :(

**Comment by lysnikolaou** [comment]
Yeah, me neither. I'm gonna continue investigating, but if someone from the @python/windows-team could enlighten us, it'd be helpful.

**Comment by lysnikolaou** [comment]
I guess the failure is unrelated to the changes of this PR. It probably has something to do with the Azure environment, since the exact same build target succeeds in the Windows Github Action, but fails with multiple compile errors on Azure.

**Comment by pablogsal** [comment]
> I guess the failure is unrelated to the changes of this PR. It probably has something to do with the Azure environment, since the exact same build target succeeds in the Windows Github Action, but fails with multiple compile errors on Azure.

But then why other PRs do not have the same problem?

**Comment by lysnikolaou** [comment]
> But then why other PRs do not have the same problem?

Unfortunately, I don't know. Will investigate more tomorrow.

**Comment by nnemkin** [comment]
The reason for the breakage is that you're indirectly including `Python-ast.h` into `Python.h` (via `pegen_interface.h`).
`Python-ast.h` defines a lot of unprefixed names (historical mistake), like `Yield` or `Set`. It should never ever be included into `Python.h` (or any other header for that matter). Consider using `struct _mod` instead of `mod_ty` in header files.

**Comment by pablogsal** [comment]
> Consider using `struct _mod` instead of `mod_ty` in header files.

Or split the header file into a private version and a public one

**Comment by lysnikolaou** [comment]
> It should never ever be included into `Python.h` (or any other header for that matter). Consider using `struct _mod` instead of `mod_ty` in header files.

And I was wondering why in `pythonrun.h` `struct _mod *` is used instead of `mod_ty`.

**Review by pablogsal — APPROVED** [review]

File: Include/pegen_interface.h [hunk 1]
```diff
@@ -4,43 +4,49 @@
 extern "C" {
```
Comment (pablogsal) [inline Include/pegen_interface.h]: At this point, we could rename this header to `parser_interfance.h` so the pegen name does not leak to the public stuff. What do you think?

Comment (lysnikolaou) [reply Include/pegen_interface.h]: Very good point! Done in https://github.com/python/cpython/pull/21016/commits/fd6bbe2a9a20eded451e0d2119e0cd294b8cca86.

**Pull request merged**